### PR TITLE
refactor: use @Immutable, add `ImmutableWrapper` and `ImmutableListWrapper` to make @Composable functions skippable

### PR DESCRIPTION
**Pull request by hoc081098**
- Add `@Immutable` to Ui state classes.
- Use `ImmutableWrapper<T>` instead of `T` when T is unstable (eg. domain model) in composable functions to make them skippable (ImmutableWrapper is @JvmInline value class -> no performance overhead).
- Use `ImmutableListWrapper<T>` instead of `List<T>` when T is unstable (eg. domain model) in composable functions to make them skippable (ImmutableListWrapper is @JvmInline value class -> no performance overhead).
- Use `kotlinx.collections.immutable.ImmutableList<T>`/`kotlinx.collections.immutable.ImmutableSet<T>` instead of `List<T>`/`Set<T>` when T is stable.

# CURRENT

```
gradle app:assembleProdRelease -PenableComposeCompilerReports=true
```

### BOOKMARKS

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun BookmarksScreen(
  unstable feedState: NewsFeedUiState
  stable removeFromBookmarks: Function1<String, Unit>
  stable onTopicClick: Function1<String, Unit>
  stable modifier: Modifier? = @static Companion
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun BookmarksGrid(
  unstable feedState: NewsFeedUiState
  stable removeFromBookmarks: Function1<String, Unit>
  stable onTopicClick: Function1<String, Unit>
  stable modifier: Modifier? = @static Companion
)
```

### FOR YOU

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun ForYouScreen(
  stable isSyncing: Boolean
  onboardingUiState: OnboardingUiState
  unstable feedState: NewsFeedUiState
  stable onTopicCheckedChanged: Function2<String, Boolean, Unit>
  stable onTopicClick: Function1<String, Unit>
  stable saveFollowedTopics: Function0<Unit>
  stable onNewsResourcesCheckedChanged: Function2<String, Boolean, Unit>
  stable modifier: Modifier? = @static Companion
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun TopicSelection(
  unstable onboardingUiState: Shown
  stable onTopicCheckedChanged: Function2<String, Boolean, Unit>
  stable modifier: Modifier? = @static Companion
)
```

### INTERESTS

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun TopicsTabContent(
  unstable topics: List<FollowableTopic>
  stable onTopicClick: Function1<String, Unit>
  stable onFollowButtonClick: Function2<String, Boolean, Unit>
  stable modifier: Modifier? = @static Companion
)
```

### TOPIC

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun TopicToolbar(
  unstable uiState: FollowableTopic
  stable modifier: Modifier? = @static Companion
  stable onBackClick: Function0<Unit>? = @static {

}
```

### APP

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun NiaApp(
  stable windowSizeClass: WindowSizeClass
  unstable networkMonitor: NetworkMonitor
  stable appState: NiaAppState? = @dynamic rememberNiaAppState(windowSizeClass, networkMonitor, null, null, $composer, 0b01000000 or 0b1110 and $dirty, 0b1100)
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun NiaNavRail(
  unstable destinations: List<TopLevelDestination>
  stable onNavigateToDestination: Function1<TopLevelDestination, Unit>
  unstable currentDestination: NavDestination?
  stable modifier: Modifier? = @static Companion
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun NiaBottomBar(
  unstable destinations: List<TopLevelDestination>
  stable onNavigateToDestination: Function1<TopLevelDestination, Unit>
  unstable currentDestination: NavDestination?
  stable modifier: Modifier? = @static Companion
)
```

### CORE UI

```
restartable scheme("[androidx.compose.ui.UiComposable]") fun NewsResourceCardExpanded(
  unstable userNewsResource: UserNewsResource
  stable isBookmarked: Boolean
  stable onToggleBookmark: Function0<Unit>
  stable onClick: Function0<Unit>
  stable onTopicClick: Function1<String, Unit>
  stable modifier: Modifier? = @static Companion
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun NewsResourceMetaData(
  unstable publishDate: Instant
  stable resourceType: NewsResourceType
)
restartable scheme("[androidx.compose.ui.UiComposable]") fun NewsResourceTopics(
  unstable topics: List<FollowableTopic>
  stable onTopicClick: Function1<String, Unit>
  stable modifier: Modifier? = @static Companion
)
```

# AFTER

There are some unskippable functions left: `NiaApp`, `NiaNavRail`, `NiaBottomBar`

### CODE REVIEW

File: core/ui/src/main/java/com/google/samples/apps/nowinandroid/core/ui/NewsResourceCard.kt [hunk 1]
```diff
@@ -66,7 +66,7 @@ import kotlinx.datetime.Instant
 import kotlinx.datetime.toJavaInstant
 import java.time.ZoneId
 import java.time.format.DateTimeFormatter
-import java.util.Locale
+import java.util.*
```
Comment (hoc081098) [inline core/ui/src/main/java/com/google/samples/apps/nowinandroid/core/ui/NewsResourceCard.kt:69]:
```suggestion
import java.util.Locale
```

**Comment by dungngminh** [comment]
Nice 🤤

**Review by mmoczkowski — COMMENTED** [review]
Interesting, but I'm not sure I like the increased complexity and decreased readability. Let's ask the compose expert. @simona-anomis would you be so kind and check this PR?

File: feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksViewModel.kt [hunk 2]
```diff
@@ -40,7 +42,8 @@ class BookmarksViewModel @Inject constructor(
 
     val feedUiState: StateFlow<NewsFeedUiState> =
         userNewsResourceRepository.observeAllBookmarked()
-            .map<List<UserNewsResource>, NewsFeedUiState>(NewsFeedUiState::Success)
+            .map { newsResources -> newsResources.filter(UserNewsResource::isSaved) } // Only show bookmarked news resources.
```
Comment (hoc081098) [inline feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksViewModel.kt:45]:
```suggestion
```

Comment (hoangchungk53qx1) [reply feature/bookmarks/src/main/java/com/google/samples/apps/nowinandroid/feature/bookmarks/BookmarksViewModel.kt]: A really good PR, it makes compose more stable when using Immutable

**Review by bentrengrove — CHANGES_REQUESTED** [review]
Thank you for the contribution and effort. Unfortunately, we can't merge a large PR like this. To land a performance change like this you would need to show the changes actually improved performance and were worth the increase in code complexity. You can see an example of this here, https://github.com/android/nowinandroid/pull/380

You would also need to split the PR up into smaller changes with an accompanied benchmark test for each change showing the before and after performance change please.

**Comment by mmoczkowski** [comment]
Thank you for your hard work. As Ben says, we can't merge it as is, but you can still create smaller PRs with documented performance improvements.

**Pull request closed without merging**